source/fuzz/transformation_composite_insert.cpp

```diff
@@ -0,0 +1,225 @@
+
+  if (!IsCompositeInstructionSupported(ir_context, composite)) {
+    return false;
+  }
+
+  // |message_.index| must refer to a valid index.
+  auto component_to_be_replaced_type_id = fuzzerutil::WalkCompositeTypeIndices(
+      ir_context, composite->type_id(), message_.index());
+  if (component_to_be_replaced_type_id == 0) {
+    return false;
+  }
+  // The instruction having the id of |message_.object_id| must be valid.
+  auto object_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.object_id());
+  if (object_instruction == nullptr) {
+    return false;
+  }
+  if (object_instruction->type_id() == 0) {
+    return false;
+  }
+
+  // We ignore pointers for now. Consider adding support for pointer types.
+  auto object_instruction_type =
+      ir_context->get_type_mgr()->GetType(object_instruction->type_id());
+  // No unused variables in release mode (to keep compilers happy).
+  (void)object_instruction_type;
+  if (object_instruction_type->AsPointer() != nullptr) {
+    return false;
+  }
+
+  // The type id of the object having |message_.object_id| and the type id of
+  // the component of the composite at index |message_.index| must be the same.
+  if (component_to_be_replaced_type_id != object_instruction->type_id()) {
+    return false;
+  }
+
+  // |message_.instruction_to_insert_before| must refer to a valid
+  // instruction.
+  auto instruction_to_insert_before =
+      FindInstruction(message_.instruction_to_insert_before(), ir_context);
+  if (instruction_to_insert_before == nullptr) {
+    return false;
+  }
+
+  // |message_.composite_id| and |message_.object_id| must be available before
+  // the |message_.instruction_to_insert_before|.
+  if (!fuzzerutil::IdIsAvailableBeforeInstruction(
+          ir_context, instruction_to_insert_before, message_.composite_id())) {
+    return false;
+  }
+  if (!fuzzerutil::IdIsAvailableBeforeInstruction(
+          ir_context, instruction_to_insert_before, message_.object_id())) {
+    return false;
+  }
+
+  // Check whether it is legitimate to insert an OpCompositeInsert before this
+  // instruction.
+  return fuzzerutil::CanInsertOpcodeBeforeInstruction(
+      SpvOpCompositeInsert, instruction_to_insert_before);
+}
+
+void TransformationCompositeInsert::Apply(
+    opt::IRContext* ir_context,
+    TransformationContext* transformation_context) const {
+  // |message_.struct_fresh_id| must be fresh.
+  assert(fuzzerutil::IsFreshId(ir_context, message_.fresh_id()) &&
+         "|message_.fresh_id| must be fresh");
+
+  std::vector<uint32_t> index =
+      fuzzerutil::RepeatedFieldToVector(message_.index());
+  opt::Instruction::OperandList in_operands;
+  in_operands.push_back({SPV_OPERAND_TYPE_ID, {message_.object_id()}});
+  in_operands.push_back({SPV_OPERAND_TYPE_ID, {message_.composite_id()}});
+  for (uint32_t i = 0; i < index.size(); i++) {
+    in_operands.push_back({SPV_OPERAND_TYPE_LITERAL_INTEGER, {index[i]}});
+  }
```

Comment:
Consider using a range-for loop here (`for (auto i : index)`).

Reply:
I've tried it and it no longer preserved the increasing order of index i (0->1->2 ...). I have left it as it is.

Reply:
I can be a little bit confusing but you would have to use `i` instead of `index[i]` once you've replaced the loop. Range-for loop iterates through the values in the vector, not indices into it.

Reply:
Yes, you're right, I missed that. Thanks.

Reply:
:+1: